# Vehicles Analytics <a id="title"></a>

# Contents <a id="back"></a>

* [Introduction](#intro)
* [Stage 1. Data overview](#data_review)
    * [Conclusions](#data_review_conclusions)
* [Stage 2. Data preprocessing](#data_preprocessing)
    * [2.1 Header style](#header_style)
    * [2.2 Missing values](#missing_values)
    * [2.3 Duplicates](#duplicates)
    * [2.4 Adding new columns](#adding_new_columns) 
    * [2.5 Conclusions](#data_preprocessing_conclusions)
* [Stage 3. Testing the hypotheses](#hypotheses)
    * [3.1 Hypothesis 1: car price dependence on odometer readings and car age](#odometer)
    * [3.2 Hypothesis 2: car types popularity](#type)
    * [3.3 Hypothesis 3: car condition](#condition)
    * [3.4 Hypothesis 4: transmission impact on market distribution and prices](#transmission)
* [Findings](#end)

## Introduction <a id='intro'></a>
In this project, we compare US second hand car pricing. We study how different second hand car characteristics affect its price.

### Goal: 
Test three hypotheses:
1. Second hand cars with high odometer are cheapper then same cars with lower odometer. "Older" cars has higher odometer. 
2. SUV cars are more popular then other type.
3. Most of the cars in secon hand market are in an excellent and a good condition. 
3. There are more automatic transmission cars in the market, automatic to manual transmission relation of pickups is different and there are more manual transmission cars of this type.

### Stages  
Project consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Testing the hypotheses
 
[Back to Contents](#back)

## Stage 1. Data overview <a id='data_review'></a>

In [73]:
# Plotly install
!pip install plotly

In [74]:
# Importing pandas
import pandas as pd

# Importing plotly.express
import plotly.express as px

In [75]:
# Reading the file and storing it to df
try:
    df = pd.read_csv('C:/Users/count/Project_sprint_6/vehicles_us.csv')
      
except:
    df = pd.read_csv('https://practicum-content.s3.us-west-1.amazonaws.com/datasets/vehicles_us.csv') 

In [76]:
# Obtaining the first 10 rows from the df table
df.head(10) 

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
5  14990      2014.0    chrysler 300  excellent        6.0  gas   57954.0   
6  12990      2015.0    toyota camry  excellent        4.0  gas   79212.0   
7  15990      2013.0     honda pilot  excellent        6.0  gas  109473.0   
8  11500      2012.0     kia sorento  excellent        4.0  gas  104174.0   
9   9200      2008.0     honda pilot  excellent        NaN  gas  147191.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28  
5    automatic   sedan       black     1.0  2018-06-20           15  
6    automatic   sedan       white     NaN  2018-12-27           73  
7    automatic     SUV       black     1.0  2019-01-07           68  
8    automatic     SUV         NaN     1.0  2018-07-16           19  
9    automatic     SUV        blue     1.0  2019-02-15           17

In [77]:
# Obtaining general information about the data in df
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


The table contains thirteen columns. They store different data types: `object`, 'int', 'float'.

According to the documentation:
- `'price'` — car price
- `'model_year'` — car model year
- `'model'` — car model
- `'condition'` - car condition
- `'cylinders'` — cylinders number
- `'fuel'` — fuel type
- `'odometer'` — total distance travelled by the car
- `'transmission'` — car transmission type
- `'type'` — car type
- `'paint_color'` - paint color
- `'is_4wd'` — 4wd car or not
- `'date_posted'` 
- `'days_listed'` 

There are no issues with style in the column names.

The number of column values is different. This means the data contains missing values.

### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores car marketing data. Columns describe car caracteristics: basic (model and model year, car type, number of cylinders, transmission type, paint color, 4wd car or not) and custom (condition, price, odometer). Also there are date posted and number of days listed columns.

It's clear that the data is sufficient to test the hypotheses. However, there are missing values.

To move forward, we need to preprocess the data.

[Back to Contents](#back)

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers (not needed) and deal with the missing values. Then, check whether there are duplicates in the data.

### Missing values <a id='missing_values'></a>

In [78]:
# Calculating missing values
df.isnull().sum() 

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

Not all missing values affect the research. For instance, the missing values in `cylinders` and `is_4wd` are not critical. It can be simply replaced with clear markers.

But missing values in `'odometer'` can affect the comparison of car advertisement data set. In real life, it would be useful to learn the reasons why the data is missing and try to make up for them. But we do not have that opportunity in this project. So we have to:
* Fill in these missing values with markers
* Evaluate how much the missing values may affect our computations

In [79]:
# Replace the missing values in `'model_year'`, `'cylinders'`, 'paint_color'` and `'is_4wd'` with the string `'unknown'`. 
columns_to_replace = ['model_year', 'cylinders', 'paint_color', 'is_4wd']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown') 

# For futher analyses required to keep 'odometer' with float dtype.
# Replace the missing values in `'odometer'` with `'-1'`. 
data_to_replace = ['odometer']
for column in data_to_replace:
    df[column] = df[column].fillna('-1')  

In [80]:
# Counting missing values again
df.isna().sum() 

price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64

### Duplicates <a id='duplicates'></a>

There are no obvious duplicates:

In [81]:
# Counting clear duplicates
df.duplicated().sum()

0

### Implicit duplicates <a id='duplicates'></a>

In [82]:
# Viewing unique model names
df['model'].sort_values().unique() 

array(['acura tl', 'bmw x5', 'buick enclave', 'cadillac escalade',
       'chevrolet camaro', 'chevrolet camaro lt coupe 2d',
       'chevrolet colorado', 'chevrolet corvette', 'chevrolet cruze',
       'chevrolet equinox', 'chevrolet impala', 'chevrolet malibu',
       'chevrolet silverado', 'chevrolet silverado 1500',
       'chevrolet silverado 1500 crew', 'chevrolet silverado 2500hd',
       'chevrolet silverado 3500hd', 'chevrolet suburban',
       'chevrolet tahoe', 'chevrolet trailblazer', 'chevrolet traverse',
       'chrysler 200', 'chrysler 300', 'chrysler town & country',
       'dodge charger', 'dodge dakota', 'dodge grand caravan',
       'ford econoline', 'ford edge', 'ford escape', 'ford expedition',
       'ford explorer', 'ford f-150', 'ford f-250', 'ford f-250 sd',
       'ford f-250 super duty', 'ford f-350 sd', 'ford f150',
       'ford f150 supercrew cab xlt', 'ford f250', 'ford f250 super duty',
       'ford f350', 'ford f350 super duty', 'ford focus', 'ford focus

There are several implicit duplicates to be deleted. For example, 'ford f-150' and 'ford f150'.

In [83]:
# Function for replacing implicit duplicates
def replace_wrong_models (wrong_models, correct_models):
    for wrong_model in wrong_models:
        df['model'] = df ['model'].replace(wrong_model, correct_model) 

In [84]:
# Removing implicit duplicates. 
# Removing specific details (e.g., trim level) and keeping only the base model name. 
wrong_models = ['chevrolet camaro', 'chevrolet camaro lt coupe 2d']
correct_model = 'chevrolet camaro' 
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['ford f150', 'ford f150 supercrew cab xlt']
correct_model = 'ford f-150'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['chevrolet silverado 1500', 'chevrolet silverado 1500 crew', 'chevrolet silverado 2500hd', 
                'chevrolet silverado 3500hd']
correct_model = 'chevrolet silverado' 
replace_wrong_models (wrong_models, correct_model)

wrong_models = ['ford f-250 sd', 'ford f-250 super duty', 'ford f250', 'ford f250 super duty']
correct_model = 'ford f-250'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['ford f-350 sd', 'ford f350', 'ford f350 super duty']
correct_model = 'ford f-350'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['ford focus', 'ford focus se']
correct_model = 'ford focus'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['ford fusion', 'ford fusion se']
correct_model = 'ford fusion'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['ford mustang', 'ford mustang gt coupe 2d']
correct_model = 'ford mustang'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['gmc sierra 1500', 'gmc sierra 2500hd']
correct_model = 'gmc sierra'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['honda civic', 'honda civic lx']
correct_model = 'honda civic'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['jeep grand cherokee', 'jeep grand cherokee laredo']
correct_model = 'jeep grand cherokee'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['jeep wrangler', 'jeep wrangler unlimited']
correct_model = 'jeep wrangler'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['nissan frontier', 'nissan frontier crew cab sv']
correct_model = 'nissan frontier'
replace_wrong_models (wrong_models, correct_model) 

wrong_models = ['toyota camry', 'toyota camry le']
correct_model = 'toyota camry'
replace_wrong_models (wrong_models, correct_model)    


In [85]:
# Checking for implicit duplicates again
sorted(df['model'].unique()) 

['acura tl',
 'bmw x5',
 'buick enclave',
 'cadillac escalade',
 'chevrolet camaro',
 'chevrolet colorado',
 'chevrolet corvette',
 'chevrolet cruze',
 'chevrolet equinox',
 'chevrolet impala',
 'chevrolet malibu',
 'chevrolet silverado',
 'chevrolet suburban',
 'chevrolet tahoe',
 'chevrolet trailblazer',
 'chevrolet traverse',
 'chrysler 200',
 'chrysler 300',
 'chrysler town & country',
 'dodge charger',
 'dodge dakota',
 'dodge grand caravan',
 'ford econoline',
 'ford edge',
 'ford escape',
 'ford expedition',
 'ford explorer',
 'ford f-150',
 'ford f-250',
 'ford f-350',
 'ford focus',
 'ford fusion',
 'ford mustang',
 'ford ranger',
 'ford taurus',
 'gmc acadia',
 'gmc sierra',
 'gmc yukon',
 'honda accord',
 'honda civic',
 'honda cr-v',
 'honda odyssey',
 'honda pilot',
 'hyundai elantra',
 'hyundai santa fe',
 'hyundai sonata',
 'jeep cherokee',
 'jeep grand cherokee',
 'jeep liberty',
 'jeep wrangler',
 'kia sorento',
 'kia soul',
 'mercedes-benz benze sprinter 2500',
 'niss

In [86]:
# Removing implicit duplicates in 'type' column. 
# Checking for implicit duplicates
sorted(df['type'].unique())

['SUV',
 'bus',
 'convertible',
 'coupe',
 'hatchback',
 'mini-van',
 'offroad',
 'other',
 'pickup',
 'sedan',
 'truck',
 'van',
 'wagon']

In [87]:
# Function for replacing implicit duplicates in 'type' column
def replace_wrong_type (wrong_types, correct_type):
    for wrong_type in wrong_types:
        df['type'] = df ['type'].replace(wrong_types, correct_type) 

In [88]:
# Removing implicit duplicates. 
wrong_types = ['truck', 'pickup']
correct_type = 'pickup' 
replace_wrong_type (wrong_types, correct_type) 

In [89]:
# Checking for implicit duplicates again
sorted(df['type'].unique())

['SUV',
 'bus',
 'convertible',
 'coupe',
 'hatchback',
 'mini-van',
 'offroad',
 'other',
 'pickup',
 'sedan',
 'van',
 'wagon']

For more convinient table structure `'model'` column should be moved to the beginning of the table

In [90]:
# Show all columns
df.columns

Index(['price', 'model_year', 'model', 'condition', 'cylinders', 'fuel',
       'odometer', 'transmission', 'type', 'paint_color', 'is_4wd',
       'date_posted', 'days_listed'],
      dtype='object')

In [91]:
# Move `'model'` column to the beginning of the table
df = df[['model', 'price', 'model_year', 'condition', 'cylinders', 'fuel',
       'odometer', 'transmission', 'type', 'paint_color', 'is_4wd',
       'date_posted', 'days_listed']]

### Adding new columns <a id='adding_new_columns'></a>

* Car age calculation

In [92]:
# Convert 'date_posted' column to datetime format
df['date_posted'] = pd.to_datetime(df['date_posted'])

# Extract publication year
df['year_posted'] = df['date_posted'].dt.year

# Convert 'year_posted' and 'model_year' columns to numeric type
df['year_posted'] = pd.to_numeric(df['year_posted'], errors='coerce')
df['model_year'] = pd.to_numeric(df['model_year'], errors='coerce')

# Calculate the age of the car and adjust for days listed
df['car_age'] = df['year_posted'] - df['model_year']
df['car_age'] -= df['days_listed'] / 365
df['car_age'] = df['car_age'].round(2)

print(df.head(10))

            model  price  model_year  condition cylinders fuel  odometer  \
0          bmw x5   9400      2011.0       good       6.0  gas  145000.0   
1      ford f-150  25500         NaN       good       6.0  gas   88705.0   
2  hyundai sonata   5500      2013.0   like new       4.0  gas  110000.0   
3      ford f-150   1500      2003.0       fair       8.0  gas        -1   
4    chrysler 200  14900      2017.0  excellent       4.0  gas   80903.0   
5    chrysler 300  14990      2014.0  excellent       6.0  gas   57954.0   
6    toyota camry  12990      2015.0  excellent       4.0  gas   79212.0   
7     honda pilot  15990      2013.0  excellent       6.0  gas  109473.0   
8     kia sorento  11500      2012.0  excellent       4.0  gas  104174.0   
9     honda pilot   9200      2008.0  excellent   unknown  gas  147191.0   

  transmission    type paint_color   is_4wd date_posted  days_listed  \
0    automatic     SUV     unknown      1.0  2018-06-23           19   
1    automatic  pic

Car age rows with missing value should be deleted from the sample. It is important characteristics for car price analytics.

After deleting rows required to do re-indexing.

In [93]:
# Calculating missing values
df.isnull().sum() 

model              0
price              0
model_year      3619
condition          0
cylinders          0
fuel               0
odometer           0
transmission       0
type               0
paint_color        0
is_4wd             0
date_posted        0
days_listed        0
year_posted        0
car_age         3619
dtype: int64

In [94]:
# Drop rows with NaN values only in the 'car_age' column
df = df.dropna(subset=['car_age'])

In [95]:
# Reset index
df = df.reset_index(drop=True)
df.head(10)

model  price  model_year  condition cylinders fuel  odometer  \
0               bmw x5   9400      2011.0       good       6.0  gas  145000.0   
1       hyundai sonata   5500      2013.0   like new       4.0  gas  110000.0   
2           ford f-150   1500      2003.0       fair       8.0  gas        -1   
3         chrysler 200  14900      2017.0  excellent       4.0  gas   80903.0   
4         chrysler 300  14990      2014.0  excellent       6.0  gas   57954.0   
5         toyota camry  12990      2015.0  excellent       4.0  gas   79212.0   
6          honda pilot  15990      2013.0  excellent       6.0  gas  109473.0   
7          kia sorento  11500      2012.0  excellent       4.0  gas  104174.0   
8          honda pilot   9200      2008.0  excellent   unknown  gas  147191.0   
9  chevrolet silverado  19500      2011.0  excellent       8.0  gas  128413.0   

  transmission    type paint_color   is_4wd date_posted  days_listed  \
0    automatic     SUV     unknown      1.0  2018-06-23           19   
1    automatic   sedan         red  unknown  2019-02-07           79   
2    automatic  pickup     unknown  unknown  2019-03-22            9   
3    automatic   sedan       black  unknown  2019-04-02           28   
4    automatic   sedan       black      1.0  2018-06-20           15   
5    automatic   sedan       white  unknown  2018-12-27           73   
6    automatic     SUV       black      1.0  2019-01-07           68   
7    automatic     SUV     unknown      1.0  2018-07-16           19   
8    automatic     SUV        blue      1.0  2019-02-15           17   
9    automatic  pickup       black      1.0  2018-09-17           38   

   year_posted  car_age  
0         2018     6.95  
1         2019     5.78  
2         2019    15.98  
3         2019     1.92  
4         2018     3.96  
5         2018     2.80  
6         2019     5.81  
7         2018     5.95  
8         2019    10.95  
9         2018     6.90

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected two issues with the data:

- Missing values
- Implicit duplicates

All missing values have been replaced with `'unknown'`. Missing values in `'odometer'` will affect our analysis, so it was replaced with '-1'.<br>
`'car_age'` critical for data analises, rows with missing car sge were deleted.

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to testing hypotheses. 

## Stage 3. Testing hypotheses <a id='hypotheses'></a>

### Hypothesis 1: car price dependence on odometer readings and car age <a id='odometer'></a>

According to the first hypothesis second hand cars with high odometer are cheapper then same cars with lower odometer. 

* Divide the cars into groups by odometer readings: low - medium - high - unknown.
* Count 'car's age' 
* Check and compare car price range in the groups.
* Check if "older" cars has higher odometer readings. 

Divide the cars into groups by odometer readings : low - medium - high - unknown

In [96]:
# Convert `'odometer'` to numeric data type. 
df['odometer'] = pd.to_numeric(df['odometer'])

# Define bins
bins = [-1, 0, 50000, 100000, float("inf")]

# Define labels for each group
labels = ['unknown', 'low', 'medium', 'high']

# Create a new column 'mileage' with bin labels based on the 'odometer' values
df['mileage'] = pd.cut(df['odometer'], bins=bins, labels=labels, right=False)

# Output the DataFrame
print(df.head(10))

                 model  price  model_year  condition cylinders fuel  odometer  \
0               bmw x5   9400      2011.0       good       6.0  gas  145000.0   
1       hyundai sonata   5500      2013.0   like new       4.0  gas  110000.0   
2           ford f-150   1500      2003.0       fair       8.0  gas      -1.0   
3         chrysler 200  14900      2017.0  excellent       4.0  gas   80903.0   
4         chrysler 300  14990      2014.0  excellent       6.0  gas   57954.0   
5         toyota camry  12990      2015.0  excellent       4.0  gas   79212.0   
6          honda pilot  15990      2013.0  excellent       6.0  gas  109473.0   
7          kia sorento  11500      2012.0  excellent       4.0  gas  104174.0   
8          honda pilot   9200      2008.0  excellent   unknown  gas  147191.0   
9  chevrolet silverado  19500      2011.0  excellent       8.0  gas  128413.0   

  transmission    type paint_color   is_4wd date_posted  days_listed  \
0    automatic     SUV     unknown  

Count cars by groups

In [97]:
# Group by group column (low/medium/high/unknown)
df.groupby('mileage')['model'].count()

C:\Users\count\AppData\Local\Temp\ipykernel_34468\1125328496.py:2: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



mileage
unknown     7343
low         6828
medium     10082
high       23653
Name: model, dtype: int64

In [98]:
# Add percentage share for each mileage group

# Group by 'mileage' and count occurrences
mileage_counts = df.groupby('mileage')['model'].size().reset_index(name='count')

# Calculate percentage and add it as a new column
total_count = df.shape[0]  # Total count of rows in the DataFrame
mileage_counts['percentage'] = (mileage_counts['count'] / total_count) * 100
mileage_counts['percentage'] = mileage_counts['percentage'].round(2)

print(mileage_counts)

   mileage  count  percentage
0  unknown   7343       15.33
1      low   6828       14.25
2   medium  10082       21.05
3     high  23653       49.37


C:\Users\count\AppData\Local\Temp\ipykernel_34468\2968276504.py:4: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



Most of the cars in cars marketing data are with high odometer readings. There is a need to check if there is any price dependence.

Check and compare car price range in the groups

In [99]:
# Median price by mileage
df.groupby('mileage')['price'].median()

C:\Users\count\AppData\Local\Temp\ipykernel_34468\484406294.py:2: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



mileage
unknown     9000.0
low        19600.0
medium     12995.0
high        6900.0
Name: price, dtype: float64

In [100]:
# Average price by mileage
df.groupby('mileage')['price'].mean().round(2)

C:\Users\count\AppData\Local\Temp\ipykernel_34468\1937701744.py:2: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



mileage
unknown    12101.36
low        20236.44
medium     14685.85
high        8771.06
Name: price, dtype: float64

In [101]:
# Median car age by mileage
df.groupby('mileage')['car_age'].median()

C:\Users\count\AppData\Local\Temp\ipykernel_34468\3751488616.py:2: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



mileage
unknown    6.96
low        1.90
medium     4.93
high       9.95
Name: car_age, dtype: float64

In [102]:
# Average car age by mileage
df.groupby('mileage')['car_age'].mean().round(2)

C:\Users\count\AppData\Local\Temp\ipykernel_34468\1648141690.py:2: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.



mileage
unknown     8.41
low         3.86
medium      6.49
high       10.62
Name: car_age, dtype: float64

In [103]:
# Mileage dataframe

# Group by 'mileage' and count occurrences
mileage_df1 = df.groupby('mileage')['price'].median().reset_index(name='median_price')
mileage_df2 = df.groupby('mileage')['price'].mean().round(2).reset_index(name='average_price')
mileage_df3 = df.groupby('mileage')['car_age'].median().reset_index(name='median_car_age')
mileage_df4 = df.groupby('mileage')['car_age'].mean().round(2).reset_index(name='average_car_age')

# New df creation and merge
df_merged = pd.merge(mileage_df1, mileage_df2[['mileage', 'average_price']], on='mileage', how='left')
df_merged = pd.merge(df_merged, mileage_df3[['mileage', 'median_car_age']], on='mileage', how='left')
df_merged = pd.merge(df_merged, mileage_df4[['mileage', 'average_car_age']], on='mileage', how='left')

print(df_merged)


   mileage  median_price  average_price  median_car_age  average_car_age
0  unknown        9000.0       12101.36            6.96             8.41
1      low       19600.0       20236.44            1.90             3.86
2   medium       12995.0       14685.85            4.93             6.49
3     high        6900.0        8771.06            9.95            10.62


C:\Users\count\AppData\Local\Temp\ipykernel_34468\2969971700.py:4: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

C:\Users\count\AppData\Local\Temp\ipykernel_34468\2969971700.py:5: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

C:\Users\count\AppData\Local\Temp\ipykernel_34468\2969971700.py:6: FutureWarning:

The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.

C:\Users\count\AppData\Local\Temp\ipykernel_34468\2969971700.py:7: FutureWarning:

Th

In [104]:
# Check what are the agest cars
df.sorted = df.sort_values(by='car_age', ascending=True)
# Reset indexing
df.sorted = df.sorted.reset_index(drop=True)
df.sorted.tail(10)

C:\Users\count\AppData\Local\Temp\ipykernel_34468\3260116727.py:2: UserWarning:

Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access



model  price  model_year  condition cylinders fuel  \
47896    chevrolet impala  35000      1958.0  excellent       8.0  gas   
47897    chevrolet impala  37900      1958.0       good       8.0  gas   
47898          ford f-250  23900      1955.0  excellent       6.0  gas   
47899          ford f-150  15000      1954.0  excellent   unknown  gas   
47900  chevrolet suburban  44900      1949.0       good   unknown  gas   
47901    chevrolet impala  21000      1948.0   like new       8.0  gas   
47902          ford f-150   5000      1936.0  excellent       6.0  gas   
47903          ford f-150  18000      1929.0       good       8.0  gas   
47904           gmc yukon  12995      1908.0       good       8.0  gas   
47905   cadillac escalade  12995      1908.0  excellent       8.0  gas   

       odometer transmission    type paint_color   is_4wd date_posted  \
47896    3184.0    automatic   coupe       black  unknown  2018-05-19   
47897   62799.0    automatic   coupe     unknown  unknown  2018-08-11   
47898   47180.0       manual  pickup        blue  unknown  2018-12-22   
47899    3565.0       manual  pickup       black  unknown  2019-02-16   
47900    1800.0    automatic   wagon      orange  unknown  2018-08-19   
47901    4000.0    automatic   sedan         red  unknown  2019-01-18   
47902   30000.0       manual  pickup      purple  unknown  2018-11-22   
47903      -1.0       manual   other      silver  unknown  2018-11-18   
47904  169328.0    automatic     SUV       black  unknown  2018-07-06   
47905      -1.0    automatic     SUV       white  unknown  2018-06-24   

       days_listed  year_posted  car_age  mileage  
47896           33         2018    59.91      low  
47897           10         2018    59.97   medium  
47898           61         2018    62.83      low  
47899           13         2019    64.96      low  
47900           10         2018    68.97      low  
47901           24         2019    70.93      low  
47902           10         2018    81.97      low  
47903           59         2018    88.84  unknown  
47904           34         2018   109.91     high  
47905           25         2018   109.93  unknown

In [105]:
# Plot scatter plot using Plotly Express
fig = px.scatter(df.sorted, x='car_age', y='price', color='mileage', title='Car price dependence on odometer readings and car age')
fig.show()

In [106]:
# Create a histogram using Plotly Express
fig = px.histogram(df_merged, x='mileage', y='average_car_age', color='median_price', title='Average car age by mileage', labels={'mileage': 'mileage', 'average_car_age': 'average_car_age'})

# Update y-axis name
fig.update_yaxes(title_text='average_car_age')

# Show the plot
fig.show()

**Conclusion**

Having compared cars by mileage, car age and price, we can draw the following conclusions:

1. Second hand cars with higher mileage are cheaper then the same cars with lower odometer readings. Car age is impoirtant characteristic that cannot be excluded when considering car price and mileage. Older cars mostly have higher mileage and are cheeper.

2. There are several expensive old cars with in medium milleage group. It is collectible cars.  
  
3. Market tendencion is that there are more older second hand cars with higher mileage. 

Thus, the first hypothesis has been proven true.

Important to note that there is a big 'unknown' group with unknown mileage. It is suggested to the marketing group to add this field as 'required' and fill in all the missed data.

### Hypothesis 2: car types popularity <a id='type'></a>

According to the second hypothesis SUV second hand cars are more popular then cars of other types. 

* Add column manufacturer to the table
* Check and compare car distribution by car type and manufacturer 
* Check and compare car price distribution by car type and manufacturer 

Add column `'manufacturer'` to the table

In [107]:
# Extract the first word from 'model' column to the 'manufacturer' column
df.sorted['manufacturer'] = df.sorted['model'].str.split().str[0]
df.sorted


model  price  model_year  condition cylinders fuel  \
0         chevrolet impala  21995      2019.0   like new       6.0  gas   
1      dodge grand caravan  17998      2019.0  excellent       6.0  gas   
2               ford f-150  23950      2019.0   like new   unknown  gas   
3            jeep wrangler  37995      2018.0       good       6.0  gas   
4         chevrolet camaro  25300      2018.0       good   unknown  gas   
...                    ...    ...         ...        ...       ...  ...   
47901     chevrolet impala  21000      1948.0   like new       8.0  gas   
47902           ford f-150   5000      1936.0  excellent       6.0  gas   
47903           ford f-150  18000      1929.0       good       8.0  gas   
47904            gmc yukon  12995      1908.0       good       8.0  gas   
47905    cadillac escalade  12995      1908.0  excellent       8.0  gas   

       odometer transmission    type paint_color   is_4wd date_posted  \
0          -1.0    automatic   sedan       black  unknown  2019-02-24   
1       29822.0    automatic     van       white  unknown  2019-03-12   
2        5000.0    automatic  pickup       white  unknown  2019-04-14   
3       33469.0    automatic     SUV       white      1.0  2018-05-02   
4        3568.0        other   coupe     unknown  unknown  2018-05-19   
...         ...          ...     ...         ...      ...         ...   
47901    4000.0    automatic   sedan         red  unknown  2019-01-18   
47902   30000.0       manual  pickup      purple  unknown  2018-11-22   
47903      -1.0       manual   other      silver  unknown  2018-11-18   
47904  169328.0    automatic     SUV       black  unknown  2018-07-06   
47905      -1.0    automatic     SUV       white  unknown  2018-06-24   

       days_listed  year_posted  car_age  mileage manufacturer  
0              236         2019    -0.65  unknown    chevrolet  
1              172         2019    -0.47      low        dodge  
2              163         2019    -0.45      low         ford  
3              163         2018    -0.45      low         jeep  
4              165         2018    -0.45      low    chevrolet  
...            ...          ...      ...      ...          ...  
47901           24         2019    70.93      low    chevrolet  
47902           10         2018    81.97      low         ford  
47903           59         2018    88.84  unknown         ford  
47904           34         2018   109.91     high          gmc  
47905           25         2018   109.93  unknown     cadillac  

[47906 rows x 17 columns]

Check cars quantity of each type

In [108]:
# Group by type
quantity_by_type = df.sorted.groupby(['type'])['price'].size().reset_index(name='quantity')

# Calculate percentage and add it as a new column
# Total count of rows in the DataFrame
total_count = df.sorted.shape[0] 
quantity_by_type['percentage'] = (quantity_by_type['quantity'] / total_count) * 100
quantity_by_type['percentage'] = quantity_by_type['percentage'].round(2)

quantity_by_type

type  quantity  percentage
0           SUV     11486       23.98
1           bus        24        0.05
2   convertible       419        0.87
3         coupe      2158        4.50
4     hatchback       974        2.03
5      mini-van      1075        2.24
6       offroad       204        0.43
7         other       234        0.49
8        pickup     18001       37.58
9         sedan     11306       23.60
10          van       594        1.24
11        wagon      1431        2.99

Based on group by type table the offer of 3 car types takes up more than 80% of the table. For futher analyses we continue to check and compare only these 3 car types: SUV, pickup, sedan. 

Pickup car market share is the highest - 37.58%

In [109]:
# Updated df
# Specify car types we are going to analyse
included_types = ['SUV', 'pickup', 'sedan']

# Filter df.sorted based on the included types
df_by_selected_type = df.sorted[df.sorted['type'].isin(included_types)]

# Display filtered df
df_by_selected_type.reset_index()

index              model  price  model_year  condition cylinders fuel  \
0          0   chevrolet impala  21995      2019.0   like new       6.0  gas   
1          2         ford f-150  23950      2019.0   like new   unknown  gas   
2          3      jeep wrangler  37995      2018.0       good       6.0  gas   
3          5         gmc acadia      1      2019.0  excellent       6.0  gas   
4          7         ford f-150  35495      2018.0  excellent       6.0  gas   
...      ...                ...    ...         ...        ...       ...  ...   
40788  47899         ford f-150  15000      1954.0  excellent   unknown  gas   
40789  47901   chevrolet impala  21000      1948.0   like new       8.0  gas   
40790  47902         ford f-150   5000      1936.0  excellent       6.0  gas   
40791  47904          gmc yukon  12995      1908.0       good       8.0  gas   
40792  47905  cadillac escalade  12995      1908.0  excellent       8.0  gas   

       odometer transmission    type paint_color   is_4wd date_posted  \
0          -1.0    automatic   sedan       black  unknown  2019-02-24   
1        5000.0    automatic  pickup       white  unknown  2019-04-14   
2       33469.0    automatic     SUV       white      1.0  2018-05-02   
3          -1.0    automatic     SUV      custom      1.0  2019-02-12   
4       13024.0    automatic  pickup        grey  unknown  2018-10-28   
...         ...          ...     ...         ...      ...         ...   
40788    3565.0       manual  pickup       black  unknown  2019-02-16   
40789    4000.0    automatic   sedan         red  unknown  2019-01-18   
40790   30000.0       manual  pickup      purple  unknown  2018-11-22   
40791  169328.0    automatic     SUV       black  unknown  2018-07-06   
40792      -1.0    automatic     SUV       white  unknown  2018-06-24   

       days_listed  year_posted  car_age  mileage manufacturer  
0              236         2019    -0.65  unknown    chevrolet  
1              163         2019    -0.45      low         ford  
2              163         2018    -0.45      low         jeep  
3              162         2019    -0.44  unknown          gmc  
4              157         2018    -0.43      low         ford  
...            ...          ...      ...      ...          ...  
40788           13         2019    64.96      low         ford  
40789           24         2019    70.93      low    chevrolet  
40790           10         2018    81.97      low         ford  
40791           34         2018   109.91     high          gmc  
40792           25         2018   109.93  unknown     cadillac  

[40793 rows x 18 columns]

In [110]:
# Group df by type and manufacturer
grouped_df_by_type = df_by_selected_type.groupby(['type', 'manufacturer'])['price'].count().reset_index(name='quantity')

# Sort df by type and quantity in descending order
grouped_df_by_type = grouped_df_by_type.sort_values(by=['type', 'quantity'], ascending=[True,False])

# Display sorted df
grouped_df_by_type.reset_index(drop=True)


type manufacturer  quantity
0      SUV         jeep      2577
1      SUV    chevrolet      2123
2      SUV         ford      2060
3      SUV       toyota      1107
4      SUV        honda       826
5      SUV          gmc       585
6      SUV       nissan       530
7      SUV       subaru       423
8      SUV     cadillac       281
9      SUV          kia       257
10     SUV        buick       245
11     SUV      hyundai       233
12     SUV          bmw       227
13     SUV        dodge         5
14     SUV        acura         3
15     SUV          ram         3
16     SUV     chrysler         1
17  pickup         ford      6325
18  pickup    chevrolet      4796
19  pickup          ram      3067
20  pickup          gmc      1609
21  pickup       toyota      1340
22  pickup       nissan       587
23  pickup        dodge       221
24  pickup         jeep        30
25  pickup     cadillac        12
26  pickup        honda         7
27  pickup          bmw         3
28  pickup        buick         2
29  pickup      hyundai         1
30  pickup          kia         1
31   sedan         ford      1906
32   sedan       toyota      1901
33   sedan    chevrolet      1590
34   sedan       nissan      1564
35   sedan        honda      1505
36   sedan      hyundai       791
37   sedan   volkswagen       752
38   sedan     chrysler       507
39   sedan        dodge       370
40   sedan        acura       219
41   sedan       subaru       138
42   sedan          kia        50
43   sedan          bmw         5
44   sedan         jeep         4
45   sedan     cadillac         2
46   sedan        buick         1
47   sedan          ram         1

Bar chart - Car distribution by car type and manufacturer

In [111]:
# Bar chart - Car distribution by car type
fig = px.bar(grouped_df_by_type, x='type', y='quantity', color='manufacturer', title='Car distribution by car type and manufacturer')
 
# Show the plot
fig.show()

Each car type has it's own most popular manufacturers:
* Jeep, Chevrolet and Ford in SUV car type
* Chevrolet, Ford, RAM in pickup car type
* Ford, Toyota, Chevrolet, Honda and Nissan in sedan car type.

Car price distribution by car type and manufacturer

In [112]:
# Group df by 'manufacturer' and 'type' -> average price
df_grouped_by_avg_price = df_by_selected_type.groupby(['type', 'manufacturer'])['price'].mean().round(2).reset_index(name = 'average_price')
# Sort df by average price within each group
df_sorted_by_avg_price = df_grouped_by_avg_price.sort_values(by=['type', 'average_price'], ascending=[True,False])

# Display df
df_sorted_by_avg_price

type manufacturer  average_price
3      SUV     cadillac       17435.95
5      SUV     chrysler       14995.00
14     SUV          ram       13630.00
11     SUV         jeep       12688.47
8      SUV          gmc       11937.19
4      SUV    chevrolet       11873.57
2      SUV        buick       11717.29
6      SUV        dodge       11300.00
1      SUV          bmw       11208.87
16     SUV       toyota       10960.17
15     SUV       subaru       10683.50
7      SUV         ford        9625.36
12     SUV          kia        9316.42
13     SUV       nissan        9160.97
9      SUV        honda        8857.67
10     SUV      hyundai        7897.38
0      SUV        acura        4950.00
29  pickup          ram       18341.40
28  pickup       nissan       17754.59
20  pickup    chevrolet       17408.35
23  pickup          gmc       16898.33
19  pickup     cadillac       16482.00
22  pickup         ford       15586.07
18  pickup        buick       15499.00
30  pickup       toyota       15029.12
26  pickup         jeep       10902.13
17  pickup          bmw        8966.33
24  pickup        honda        5357.00
25  pickup      hyundai        5100.00
21  pickup        dodge        4950.63
27  pickup          kia        3950.00
34   sedan     cadillac       15147.50
44   sedan          ram       12999.00
33   sedan        buick        9990.00
37   sedan        dodge        8835.63
36   sedan     chrysler        8581.07
42   sedan          kia        8166.90
45   sedan       subaru        7812.26
47   sedan   volkswagen        7470.97
40   sedan      hyundai        7285.88
46   sedan       toyota        6855.75
35   sedan    chevrolet        6849.42
43   sedan       nissan        6806.12
38   sedan         ford        6619.16
39   sedan        honda        6580.12
31   sedan        acura        6485.99
32   sedan          bmw        6140.00
41   sedan         jeep        4298.25

In [113]:
# Bar chart - Car price distribution by car type and manufacturer
fig = px.bar(df_sorted_by_avg_price, x='type', y='average_price', color='manufacturer', title='Car price distribution by car type and manufacturer', barmode='group')
 
# Show the plot
fig.show()

Pickup cars has wider range of prices depending on the manufacturer. Also most of the average prices are high. It is nessessary to see it in a separate chart.

In [114]:
# Updated df for pickup cars only
# Specify car types we are going to analyse
included_type = ['pickup']

# Filter df.sorted based on the included type
df_pickup_only = df_grouped_by_avg_price[df_grouped_by_avg_price['type'].isin(included_type)]

# Sort df_pickup_only by average price
df_sorted = df_pickup_only.sort_values(by=['average_price'], ascending=[False])

# Display df
df_sorted.reset_index()

index    type manufacturer  average_price
0      29  pickup          ram       18341.40
1      28  pickup       nissan       17754.59
2      20  pickup    chevrolet       17408.35
3      23  pickup          gmc       16898.33
4      19  pickup     cadillac       16482.00
5      22  pickup         ford       15586.07
6      18  pickup        buick       15499.00
7      30  pickup       toyota       15029.12
8      26  pickup         jeep       10902.13
9      17  pickup          bmw        8966.33
10     24  pickup        honda        5357.00
11     25  pickup      hyundai        5100.00
12     21  pickup        dodge        4950.63
13     27  pickup          kia        3950.00

In [115]:
# Bar chart - Pickup cars price distribution by manufacturer
fig = px.bar(df_sorted, x='type', y='average_price', color='manufacturer', text='manufacturer', title='Pickup type car price distribution by car type and manufacturer', barmode='group')
 
# Show the plot
fig.show()

The most expensive second hand pickup cars are RAM, Nissan, Chevrolet, GMC, Cadillac, Ford, Buick and Toyota. 

**Conclusion**

Having compared car quantity and car average price by type and manufacturer, we can draw the following conclusions:

1. Based on group by type table the offer of 3 car types takes up more than 80% of the table - SUV, pickup and sedan types. 
Pickup car market share is the highest - 37.58%

2. Each car type has it's own most popular manufacturers:
* Jeep, Chevrolet and Ford in SUV car type
* Chevrolet, Ford, RAM in pickup car type
* Ford, Toyota, Chevrolet, Honda and Nissan in sedan car type
  
3. Price distribution by manufacturer in SUV and sedan types mostly flat with some more expensive (prestigious) brands. 
Pickup type cars price distribution is different. There are 8 manufacturers with high prices - RAM, Nissan, Chevrolet, GMC, Cadillac, Ford, Buick and Toyota. 2 manufacturers with mid price and 4 with a low one.
Pickup cars are more expensive then other 2 popular types.

Thus, the second hypothesis has not been proven true. Pickup car market share is the highest, although SUV category market share is also high - 23.98%.

### Hypothesis 3: car condition <a id='condition'></a>

According to the third hypothesis most of the cars in second hand market are in an excellent and a good condition. 

* Check and compare car distribution by car type and condition
* Check and compare car price distribution by car type and condition

Check and compare car distribution by car type and condition.

In [116]:
# Group df
grouped_df_by_condition = df_by_selected_type.groupby(['type', 'condition'])['price'].count().reset_index(name='quantity')

# Sort df
grouped_df_by_condition = grouped_df_by_condition.sort_values(by=['type', 'quantity'], ascending=(True, False))

# Calculate percentages
total = grouped_df_by_condition['quantity'].sum()
grouped_df_by_condition['percentage'] = (grouped_df_by_condition['quantity'] / total) * 100
grouped_df_by_condition = grouped_df_by_condition.round(2)

# Display df
grouped_df_by_condition.reset_index(drop=True)


type  condition  quantity  percentage
0      SUV  excellent      6083       14.91
1      SUV       good      3986        9.77
2      SUV   like new      1079        2.65
3      SUV       fair       304        0.75
4      SUV        new        17        0.04
5      SUV    salvage        17        0.04
6   pickup  excellent      8017       19.65
7   pickup       good      7682       18.83
8   pickup   like new      1523        3.73
9   pickup       fair       692        1.70
10  pickup        new        55        0.13
11  pickup    salvage        32        0.08
12   sedan  excellent      5728       14.04
13   sedan       good      3997        9.80
14   sedan   like new      1222        3.00
15   sedan       fair       281        0.69
16   sedan        new        43        0.11
17   sedan    salvage        35        0.09

In [117]:
# Calculate percentage by excellent and good condition
# Specify conditions
included_conditions = ['excellent', 'good']

# Filter df based on the included conditions
df_selected_conditions = grouped_df_by_condition[grouped_df_by_condition['condition'].isin(included_conditions)]

# Calculate percentage
total_percentage = df_selected_conditions['percentage'].sum().round(2)

# Display df
print(total_percentage)

87.0


In [118]:
# Bar chart - Car distribution by car type and condition
fig = px.bar(grouped_df_by_condition, x='type', y='quantity', text='percentage', color='condition', title='Car distribution by car type and condition', barmode='group')

# Update layout to display percentages
fig.update_traces(texttemplate='%{text:.2f}%', textposition='inside')

# Show the plot
fig.show()

87% of cars in the second hand market are in exellent and good condition. 

Check and compare car price distribution by car type and condition.

In [119]:
# Group df by 'condition' and 'type' -> average price
df_grouped_by_cond = df_by_selected_type.groupby(['type', 'condition'])['price'].mean().round(2).reset_index(name = 'average_price')

# Sort df by average price within each group
df_sorted_by_cond = df_grouped_by_cond.sort_values(by=['type', 'average_price'], ascending=[True,False])

# Display df
df_sorted_by_cond.reset_index(drop=True)

type  condition  average_price
0      SUV        new       18431.82
1      SUV   like new       15331.02
2      SUV  excellent       12187.23
3      SUV       good        9086.29
4      SUV       fair        3202.80
5      SUV    salvage        2076.41
6   pickup        new       39327.95
7   pickup   like new       24332.69
8   pickup  excellent       17823.49
9   pickup       good       14651.46
10  pickup    salvage        6847.66
11  pickup       fair        4212.39
12   sedan        new       10650.58
13   sedan   like new        9113.95
14   sedan  excellent        7573.36
15   sedan       good        5856.64
16   sedan    salvage        3639.14
17   sedan       fair        2080.69

In [120]:
# Define a function to calculate the percentage within each group
def calculate_percentage(group):
    group['percentage_within_group'] = (group['average_price'] / group['average_price'].sum()) * 100
    return group

# Apply the function to each group and combine the results
df_sorted_by_cond_percentage = df_sorted_by_cond.groupby('type').apply(calculate_percentage).round(2)

# Display df
df_sorted_by_cond_percentage.reset_index(drop=True)

C:\Users\count\AppData\Local\Temp\ipykernel_34468\1869151137.py:7: DeprecationWarning:

DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.



type  condition  average_price  percentage_within_group
0      SUV        new       18431.82                    30.56
1      SUV   like new       15331.02                    25.42
2      SUV  excellent       12187.23                    20.21
3      SUV       good        9086.29                    15.06
4      SUV       fair        3202.80                     5.31
5      SUV    salvage        2076.41                     3.44
6   pickup        new       39327.95                    36.69
7   pickup   like new       24332.69                    22.70
8   pickup  excellent       17823.49                    16.63
9   pickup       good       14651.46                    13.67
10  pickup    salvage        6847.66                     6.39
11  pickup       fair        4212.39                     3.93
12   sedan        new       10650.58                    27.37
13   sedan   like new        9113.95                    23.42
14   sedan  excellent        7573.36                    19.46
15   sedan       good        5856.64                    15.05
16   sedan    salvage        3639.14                     9.35
17   sedan       fair        2080.69                     5.35

In [121]:
# Bar chart 'Price distribution by car type and condition'
fig = px.bar(df_sorted_by_cond_percentage, x='type', y='average_price', color='condition', text='percentage_within_group', title='Price distribution by car type and condition', barmode='group')

# Update layout to display percentages
fig.update_traces(texttemplate='%{text:.2f}%', textposition='inside')
 
# Show the plot
fig.show()

**Conclusion**

Having compared car quantity and car average price by type and condition, we can draw the following conclusions:

1. 87% of cars in the second hand market are in exellent and good condition. 

2. According to price distribution by car type and condition bar chart average price decreases with condition deterioration. 
* About 15-20% decrease for each degree.
* For pickup - decrease from 'new condition' to 'like new condition' is about 40%

Thus, the third hypothesis has been proven true.

### Hypothesis 4: transmission impact on market distribution and prices <a id='transmission'></a>

According to the forth hypothesis there are more automatic transmission cars in the market, 
automatic to manual transmission relation of pickups is different and there are more manual transmission cars of this type. 

* Check and compare car distribution by transmission and manufacturer 
* Check car distribution by transmission for pickups
* Check and compare car price distribution by transmission and manufacturer 

Car distribution by transmission and manufacturer 

In [122]:
# Group df by transmission and manufacturer
grouped_df_by_transmission_count = df.sorted.groupby(['transmission', 'manufacturer'])['price'].count().reset_index(name='quantity')

# Sort df
grouped_df_by_transmission_count = grouped_df_by_transmission_count.sort_values(by=['transmission', 'quantity'], ascending=(True, False))

# Display df
grouped_df_by_transmission_count.reset_index(drop=True)

transmission   manufacturer  quantity
0     automatic           ford     10634
1     automatic      chevrolet      9118
2     automatic         toyota      4740
3     automatic          honda      3012
4     automatic            ram      2838
5     automatic         nissan      2506
6     automatic           jeep      2499
7     automatic            gmc      2158
8     automatic          dodge      1081
9     automatic        hyundai      1037
10    automatic         subaru       995
11    automatic       chrysler       778
12    automatic     volkswagen       694
13    automatic            kia       464
14    automatic       cadillac       293
15    automatic          buick       255
16    automatic            bmw       240
17    automatic          acura       216
18    automatic  mercedes-benz        34
19       manual           ford       639
20       manual           jeep       445
21       manual      chevrolet       369
22       manual         toyota       253
23       manual          honda       200
24       manual         subaru       172
25       manual            ram       163
26       manual         nissan       111
27       manual     volkswagen       110
28       manual            kia        64
29       manual        hyundai        51
30       manual          dodge        27
31       manual            gmc        21
32       manual          acura         8
33       manual            bmw         5
34       manual          buick         2
35       manual       cadillac         1
36       manual       chrysler         1
37        other           ford       517
38        other      chevrolet       398
39        other         nissan       352
40        other           jeep        97
41        other            ram        86
42        other         toyota        66
43        other          dodge        50
44        other          honda        32
45        other            gmc        28
46        other            kia        16
47        other        hyundai        11
48        other         subaru        11
49        other     volkswagen         5
50        other            bmw         1
51        other       cadillac         1
52        other       chrysler         1

In [142]:
# Percentage of cars by transmission
# Group df by transmission and manufacturer
grouped_df_by_transmission_count_percentage = grouped_df_by_transmission_count.groupby(['transmission'])['quantity'].sum().reset_index(name='quantity')

# Calculate the percentage within each group
grouped_df_by_transmission_count_percentage['percentage'] = grouped_df_by_transmission_count_percentage['quantity'] / grouped_df_by_transmission_count_percentage['quantity'].sum() * 100

# Display df
grouped_df_by_transmission_count_percentage.round(2).reset_index(drop=True)

transmission  quantity  percentage
0    automatic     43592       90.99
1       manual      2642        5.51
2        other      1672        3.49

In [143]:
# Bar chart 'Car distribution by transmission and manufacturer'
fig = px.bar(grouped_df_by_transmission_count, x='transmission', y='quantity', color='manufacturer', title='Car distribution by transmission and manufacturer', barmode='group')
 
# Show the plot
fig.show()

The majority of cars of second hand car market with automatic transmission. It's almost 91%.

Car distribution by transmission for pickups

In [144]:
# Select pickups only
selected_type = ['pickup']

# Filter df.sorted based on the included type
df_pickup = df.sorted[df.sorted['type'].isin(selected_type)]

# Group df by transmission and manufacturer
grouped_pickup_df_by_transmission_count = df_pickup.groupby(['transmission', 'manufacturer'])['price'].count().reset_index(name='quantity')

# Sort df
grouped_pickup_df_by_transmission_count = grouped_pickup_df_by_transmission_count.sort_values(by=['transmission', 'quantity'], ascending=(True, False))

# Display df
grouped_pickup_df_by_transmission_count.reset_index(drop=True)

transmission manufacturer  quantity
0     automatic         ford      5797
1     automatic    chevrolet      4648
2     automatic          ram      2820
3     automatic          gmc      1576
4     automatic       toyota      1162
5     automatic       nissan       238
6     automatic        dodge       192
7     automatic         jeep        23
8     automatic     cadillac        12
9     automatic        honda         7
10    automatic          bmw         3
11    automatic        buick         2
12    automatic      hyundai         1
13       manual         ford       290
14       manual          ram       161
15       manual       toyota       152
16       manual    chevrolet        56
17       manual       nissan        35
18       manual        dodge        22
19       manual          gmc        18
20       manual         jeep         6
21       manual          kia         1
22        other       nissan       314
23        other         ford       238
24        other    chevrolet        92
25        other          ram        86
26        other       toyota        26
27        other          gmc        15
28        other        dodge         7
29        other         jeep         1

In [145]:
# Bar chart - 'Pickup cars distribution by transmission and manufacturer'
fig = px.bar(grouped_pickup_df_by_transmission_count, x='transmission', y='quantity', color='manufacturer', title='Pickup cars distribution by transmission and manufacturer', barmode='group')
 
# Show the plot
fig.show()

Price distribution by transmission and manufacturer 

In [146]:
# Group df by transmission and manufacturer
grouped_df_by_transmission_mean = df.sorted.groupby(['transmission', 'manufacturer'])['price'].mean().round(2).reset_index(name='average_price')

# Sort df
grouped_df_by_transmission_mean = grouped_df_by_transmission_mean.sort_values(by=['transmission', 'average_price'], ascending=(True, False))

# Display df
grouped_df_by_transmission_mean.reset_index(drop=True)

transmission   manufacturer  average_price
0     automatic  mercedes-benz       34900.00
1     automatic            ram       18898.86
2     automatic       cadillac       17205.75
3     automatic            gmc       15827.83
4     automatic      chevrolet       13985.12
5     automatic           jeep       13020.46
6     automatic           ford       12501.56
7     automatic          buick       11574.09
8     automatic            bmw       11026.34
9     automatic         toyota       10167.35
10    automatic            kia        9051.39
11    automatic         subaru        9000.96
12    automatic          dodge        8530.19
13    automatic       chrysler        8246.24
14    automatic         nissan        8131.25
15    automatic     volkswagen        7546.76
16    automatic          honda        7492.80
17    automatic        hyundai        7464.43
18    automatic          acura        6492.95
19       manual          buick       17400.00
20       manual      chevrolet       16323.47
21       manual            ram       13973.44
22       manual           jeep       12568.93
23       manual       cadillac        9300.00
24       manual         toyota        9299.25
25       manual           ford        8498.18
26       manual         subaru        7434.53
27       manual            kia        6302.84
28       manual            bmw        6280.00
29       manual            gmc        5970.24
30       manual     volkswagen        5402.24
31       manual          acura        5249.38
32       manual          honda        5133.46
33       manual         nissan        4897.53
34       manual       chrysler        4800.00
35       manual          dodge        4418.30
36       manual        hyundai        4127.61
37        other       cadillac       76995.00
38        other           ford       22217.59
39        other      chevrolet       21102.39
40        other         nissan       19749.37
41        other         subaru       10781.64
42        other           jeep        9204.29
43        other        hyundai        7949.18
44        other         toyota        7361.41
45        other            ram        6580.02
46        other          honda        6527.00
47        other     volkswagen        6517.00
48        other            bmw        6000.00
49        other            kia        5826.31
50        other       chrysler        4200.00
51        other            gmc        2688.46
52        other          dodge         969.24

In [147]:
# Bar chart - 'Car distribution by car type and condition'
fig = px.bar(grouped_df_by_transmission_mean, x='transmission', y='average_price', color='manufacturer', title='Price distribution by transmission and manufacturer', barmode='group')
 
# Show the plot
fig.show()

**Conclusion**

Having compared car quantity and car average price by transmission and manufacturer, we can draw the following conclusions:

1. Based on group by type table the offer of 3 car types takes up more than 80% of the table - SUV, pickup and sedan types. 
Pickup car market share is the highest - 37.58%

2. Each car type has it's own most popular manufacturers:
* Jeep, Chevrolet and Ford in SUV car type
* Chevrolet, Ford, RAM in pickup car type
* Ford, Toyota, Chevrolet, Honda and Nissan in sedan car type
  
3. Price distribution by manufacturer in SUV and sedan types mostly flat with some more expensive (prestigious) brands. 
Pickup type cars price distribution is different. There are 8 manufacturers with high prices - RAM, Nissan, Chevrolet, GMC, Cadillac, Ford, Buick and Toyota. 2 manufacturers with mid price and 4 with a low one.
Pickup cars are more expensive then other 2 popular types.

Thus, the second hypothesis has not been proven true. Pickup car market share is the highest, although SUV category market share is also high - 23.98%.

checkbox - убрать или оставить анноун
Findings